### PR TITLE
Experimental pipeline

### PR DESCRIPTION
**Pull request by iamjarvi**
Created pipeline that facilitates Continuous Integration. Features including:

- PR Triggers
- Directory Filtering that detects changes within the individual services.
- Control Logic that decides which services to test and build. (Currently 7, including base)
  - If Base service is built, all its dependent services will build.
  -  All other services can be built individually.
- Static Analysis using Sonar Cloud
- Produces artifacts for use in Deployment Stages (Future release)

### CODE REVIEW

**Review by SilentMrDave — APPROVED** [review]

File: azure-pipelines/azure-templates/build-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/build-template.yml]: be consistent about indentation of yaml sequences. e.g. inconsistent between here and the parameters just above. same for all yaml files here.

File: azure-pipelines/azure-templates/build-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/build-template.yml]: commented out code doesn't belong on an integration branch without a good reason.

File: azure-pipelines/azure-templates/build-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/build-template.yml]: values like this that will be modified for the next iteration or a different project should be set at the pipeline level and sent here via parameters/variables.

File: azure-pipelines/azure-templates/build-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/build-template.yml]: all steps should have a displayName. same for all yaml files here.
java --version seems like it was a temporary debugging step

File: azure-pipelines/azure-templates/build-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/build-template.yml]: i don't know why everyone keeps putting publishLocation in when the docs say it's not part of this task, but maybe the docs are just wrong. just double check it

File: azure-pipelines/azure-templates/docker-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/docker-template.yml]: i can't really tell what the purpose of these two separate jobs is. add some comments

File: azure-pipelines/azure-templates/docker-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/docker-template.yml]: don't you want this job to run always (`condition`/`dependsOn`), and at the end of any pipeline (even the ones that don't use this template - or do you expect all your pipelines to use this template)?

File: azure-pipelines/azure-templates/helm-deploy-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/helm-deploy-template.yml]: debug step?

File: azure-pipelines/azure-templates/helm-deploy-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/helm-deploy-template.yml]: be consistent about spacing between steps

File: azure-pipelines/azure-templates/helm-deploy-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/helm-deploy-template.yml]: double slash?

File: azure-pipelines/azure-templates/helm-deploy-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/helm-deploy-template.yml]: why the blank lines here and above the other valueFile setting?

File: azure-pipelines/azure-templates/webhooks-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/webhooks-template.yml]: the token is supposed to be secret and not on github, i'm pretty sure.
the other values that will certainly be different for the next iteration should not be in the templates, should be provided to the templates as parameters.

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: duplicated word?

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: why an extra indent from this line on?

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: there are possible edge-cases here based on [this behavior](https://git-scm.com/docs/git-config#Documentation/git-config.txt-corequotePath) which could be addressed by adding the `-z` option and making sure it's splitting on null bytes and not on newlines. or by ensuring the below logic can handle when the filenames are quoted and escaped.

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: there's repetition here we could probably extract into a function, but ok

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: there must be a way to install java 16 on the ms-hosted agents. this is ok but if we have self-hosted agents then we also want automation to set them up.

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: inconsistent indentation of |-strings

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: isn't this step superfluous if you have the above condition?

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: keep comments indented at the same level as the things they're describing

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: it's not sufficiently clear that these are sonarcloud organization, sonar project key, sonar project name.

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: steps like this are trying to do something which is better accomplished by just giving all steps, jobs, stages, etc good display names and letting the GUI display them accordingly

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: spelling *2

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: without `set -e` or some other mechanism, an early nonzero exit code will be ignored, and only the last command's exit code will be noticed. if you're confident that will work appropriately then ok

File: azure-pipelines/test-pipeline.yml
Comment (escalonn) [inline azure-pipelines/test-pipeline.yml]: this seems like something that doesn't need to be on development

**Review by escalonn — CHANGES_REQUESTED** [review]
neat. this PR should have been created 6 days ago, though, so it could track work in progress and not just suddenly exist once the work's almost done.

File: azure-pipelines/azure-templates/build-template.yml
Comment (iamjarvi) [reply azure-pipelines/azure-templates/build-template.yml]: Those are more file I need to clean up, actually.

File: azure-pipelines/test-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/test-pipeline.yml]: I left some deprecated files in by mistake. They should be removed.

File: azure-pipelines/azure-templates/build-template.yml
Comment (iamjarvi) [reply azure-pipelines/azure-templates/build-template.yml]: Azure Pipelines is putting there.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Good point, these are not specific.

File: azure-pipelines/azure-templates/build-template.yml
Comment (iamjarvi) [reply azure-pipelines/azure-templates/build-template.yml]: Deprecated. Removed.

File: azure-pipelines/azure-templates/build-template.yml
Comment (iamjarvi) [reply azure-pipelines/azure-templates/build-template.yml]: File removed.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Noted.

File: azure-pipelines/azure-templates/build-template.yml
Comment (iamjarvi) [reply azure-pipelines/azure-templates/build-template.yml]: Removed file.

File: azure-pipelines/azure-templates/docker-template.yml
Comment (iamjarvi) [reply azure-pipelines/azure-templates/docker-template.yml]: Removed. Basically what happened with these were for reference, and forgot to git ignore them.

File: azure-pipelines/azure-templates/webhooks-template.yml
Comment (iamjarvi) [reply azure-pipelines/azure-templates/webhooks-template.yml]: Updated.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Thanks!

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Fixed.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Can look into this further, but the Azure Built in option is pretty rigid.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: I think I addressed this.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: This has been removed.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Fixed!

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Checking this out!

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: I used -e for testing purposes, and I'm currently confident with  or without using it, but you've raised a good point. Is this something standard to keep in a pipeline?

Comment (escalonn) [reply azure-pipelines/main-pipeline.yml]: for some people yes, for others no.

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [reply azure-pipelines/main-pipeline.yml]: you have sudo rights to install anything you want with apt, though

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: That's a fair point. I like the readability of this. I might put this in a future release, but there are definitely more high priority work items.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: -z, according to the docs works only with --options I'm not using. I will continue to research this.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Java 16 installs on agent. Found other errors along the way and fixed them.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Added this.

File: base/test-file-will-remove
Comment (escalonn) [inline base/test-file-will-remove]: remove this. it takes me very little time to see files that shouldn't be present in the files changed tab of the PR, or in a local git diff, so that means you should catch them yourself before requesting review.

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: since we're using bash-specific syntax, this should really be a `bash` step rather than a `script` step so the pipeline is less platform-dependent. i won't block the PR for this though

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [reply azure-pipelines/main-pipeline.yml]: just in case we're not on the same page, the `-e` option to `mvn` is completely different from what i was talking about

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: we need both of these steps?
(also, have consistent spacing between steps)
(applies to above at line ~120 also)

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: 1. isn't that /jdk/ path different from the path that your jdk step above (& below) is using? or are those not related?
2. be consistent about indentation.
3. if these sonar steps are analyzing all of the services, clarify that with a brief comment
4. is it possible that variables['base'] is false, but we still have changes that we want to be analyzed with sonar but that this pipeline logic will skip?

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: 1. why only PR builds? don't you need regular CI builds to do the deployment from?
1. commented out things should be removed

File: azure-pipelines/java-16-install.sh
Comment (escalonn) [inline azure-pipelines/java-16-install.sh:5]: it'd be nice if there were some comments here or out on the step letting people know what considerations went into this implementation, links to docs you used, how they can go about updating it to the next version of java 16, etc. that should be your frame of mind anytime you implement something non-obvious.
we're running out of time so i won't block it based on this.

File: azure-pipelines/azure-templates/helm-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/helm-template.yml]: why so many random newlines in this & other files? it's sloppy. not blocking.

File: azure-pipelines/azure-templates/docker-template.yml
Comment (escalonn) [reply azure-pipelines/azure-templates/docker-template.yml]: ^ previous comment appears not addressed

**Review by escalonn — CHANGES_REQUESTED** [review]

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: This is for testing. It's difficult to read logs when multiple pipelines run from one push.

File: azure-pipelines/azure-templates/helm-template.yml
Comment (iamjarvi) [reply azure-pipelines/azure-templates/helm-template.yml]: Removed.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: It is intended to have CI Triggers for each of the three significant branches, plus the pipeline-test.

File: base/test-file-will-remove
Comment (iamjarvi) [reply base/test-file-will-remove]: I need to be able to have something for my pipeline to detect? Is there another way to introduce a change within a directory to trigger the main functionality of the pipeline, which is to detect changes in a directory?

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Got it.

File: base/test-file-will-remove
Comment (iamjarvi) [reply base/test-file-will-remove]: Removed.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: I believe I've taken care of all of this.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Put back in.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Both of these steps are necessary. One is downloading the JDK. One is installing it with the tool. Clarified by naming, and spaced the tasks appropriately.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: I was on the wrong page. Definitely misinterpreted that, and I understand now.

File: base/test-file-will-remove
Comment (escalonn) [reply base/test-file-will-remove]: that makes sense, but by the time it goes to review the file should be removed.

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: spelling. also, this line and the next line look like two different sentences at first glance.
(not blocking)

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: keep indentation consistent (not blocking)

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: if you're sure this is a good setting for sonar.java.binaries, then ok

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: displayname consistency, the rest don't use "ing". (not blocking)

File: azure-pipelines/azure-templates/webhooks-template.yml
Comment (escalonn) [inline azure-pipelines/azure-templates/webhooks-template.yml:12]: inconsistency in yaml indentation (not blocking)

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml]: inconsistency in yaml indentation here and other places in this file (not blocking)

**Review by escalonn — CHANGES_REQUESTED** [review]
is it correct that the pipeline doesn't ever reference the webhooks template or helm template?

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Changed.

File: azure-pipelines/azure-templates/webhooks-template.yml
Comment (iamjarvi) [reply azure-pipelines/azure-templates/webhooks-template.yml:12]: Last push should've solved this.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Last push should've solved this.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Fixed with sonarcloud rework.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: Fixed.

**Comment by iamjarvi** [comment]
The templates aren't in this PR, because, I'll be testing them in the devops_pipeline branch, once this and Eric's PRs were approved.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml]: I'm still sort of unsure what's wrong with the indentation here. But I refactored the whole .yml file so it may have been fixed.

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [inline azure-pipelines/main-pipeline.yml:25]: indicate via comment that this variable is the prefix to a set of project names, to be combined with the service names defined below.

File: base/test
Comment (escalonn) [inline base/test]: remove

**Review by escalonn — APPROVED** [review]

File: azure-pipelines/main-pipeline.yml
Comment (escalonn) [reply azure-pipelines/main-pipeline.yml]: the properties value was indented by 4 extra spaces instead of the 2 used elsewhere. fixed during the refactor.

**Review by escalonn — CHANGES_REQUESTED** [review]
oops, that was meant to be a reject

**Review by SilentMrDave — APPROVED** [review]

File: base/test
Comment (iamjarvi) [reply base/test]: Removed! I WILL remember this, next time.

File: azure-pipelines/main-pipeline.yml
Comment (iamjarvi) [reply azure-pipelines/main-pipeline.yml:25]: I changed the variable name, does that work?

**Review by escalonn — APPROVED** [review]

**Pull request merged**
